pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+	/// message.
+	///
+	/// NOTE - this gets increased by the `expected_proof_count` for a set of
+	/// routers (see `get_expected_proof_count`) every time a new identical
+	/// message is submitted.
+	pub expected_proof_count: u32,
+}
+
+/// Type that holds the information needed for inbound proof entries.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct ProofEntry<T: Config> {
+	/// The session ID for this entry.
+	pub session_id: T::SessionId,
+
+	/// The number of proofs received for a particular message.
+	///
+	/// NOTE - this gets increased by 1 every time a new identical message is
+	/// submitted.
+	pub current_count: u32,
+}
+
+impl<T: Config> ProofEntry<T> {
+	/// Returns `true` if all the following conditions are true:
+	/// - the session IDs match
+	/// - the `current_count` is greater than 0
+	pub fn has_valid_vote_for_session(&self, session_id: T::SessionId) -> bool {
+		self.session_id == session_id && self.current_count > 0
+	}
+}
+
+/// Type used when storing inbound message information.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub enum InboundEntry<T: Config> {
+	Message(MessageEntry<T>),
+	Proof(ProofEntry<T>),
+}
+
+impl<T: Config> From<(&InboundProcessingInfo<T>, T::Message)> for InboundEntry<T> {
+	fn from((inbound_processing_info, message): (&InboundProcessingInfo<T>, T::Message)) -> Self {
+		match message.get_message_proof() {
+			None => InboundEntry::Message(MessageEntry {
+				session_id: inbound_processing_info.current_session_id.clone(),
+				domain_address: inbound_processing_info.domain_address.clone(),
+				message,
+				expected_proof_count: inbound_processing_info.expected_proof_count_per_message,
+			}),
+			Some(_) => InboundEntry::Proof(ProofEntry {
+				session_id: inbound_processing_info.current_session_id,
+				current_count: 1,
+			}),
+		}
+	}
+}
+
+impl<T: Config> From<MessageEntry<T>> for InboundEntry<T> {
+	fn from(message_entry: MessageEntry<T>) -> Self {
+		Self::Message(message_entry)
+	}
+}
+
+impl<T: Config> From<ProofEntry<T>> for InboundEntry<T> {
+	fn from(proof_entry: ProofEntry<T>) -> Self {
+		Self::Proof(proof_entry)
+	}
+}
+
+impl<T: Config> InboundEntry<T> {
+	/// Creates a new `InboundEntry` based on the information provided.
+	///
+	/// If a session change is detected or if the updated counts reach 0, it
+	/// means that a new entry is no longer required, otherwise, the counts are
+	/// decreased accordingly, based on the entry type.
+	pub fn create_post_voting_entry(
+		inbound_entry: &InboundEntry<T>,
+		inbound_processing_info: &InboundProcessingInfo<T>,
+	) -> Result<Option<Self>, DispatchError> {
+		match inbound_entry {
+			InboundEntry::Message(message_entry) => {
+				if message_entry.session_id != inbound_processing_info.current_session_id {
+					return Ok(None);
+				}
+
```

Comment:
At this stage, this can never happen, right?

Reply:
Yes, same as the case we discussed for `execute_post_voting_dispatch`. It can only happen if the function is called without confirming that a message is ready for processing. I think we can skip these checks. Same goes for the other note below.

Reply:
Sorry, Cosmin, maybe I am pointing you to a bad state. Now I think those checks could be in charge of removing the message entry. Could that be?

Reply:
No, those were the ones related to the post voting updates, and in those cases, all pending inbound entries are valid.

Reply:
Is this still a relevant issue?

Reply:
No, this was related to removing entries if a session change is detected, however, that won't be the case in this scenario since this logic only gets called if all storage entries are present for the current session.